src/raftstore/store/fsm/apply.rs

```diff
@@ -2570,23 +2608,26 @@ impl ApplyFsm {
         if self.delegate.pending_remove || self.delegate.stopped {
             return;
         }
+        if apply_ctx.timer.is_none() {
+            apply_ctx.timer = Some(SlowTimer::new());
+        }
 
-        // Persists any pending changes of this region.
-        let region_id = self.delegate.id();
-        if apply_ctx
-            .apply_res
-            .iter()
-            .any(|res| res.region_id == region_id)
-        {
-            apply_ctx.flush();
+        if apply_ctx.raft_wb.is_none() {
+            apply_ctx.raft_wb = Some(WriteBatch::with_capacity(DEFAULT_RAFT_WB_SIZE));
         }
+        self.delegate.write_apply_state(apply_ctx.raft_wb_mut());
```

Comment:
What if snapshot is requested before a snapshot is applied? For example, a leader requests a snapshot, but the messages buffered in channel for a long time that it step down and start accepting a snapshot from a new leader. So apply state will be written in both raftstore and apply threads.